### PR TITLE
Flow ebos fix fip

### LINKED ISSUE #988: FIP output problem in flow_ebos
The flow_ebos program suffers from the same problem that was fixed in #987. This must be addressed.

### PR DESCRIPTION
**Pull request by andlaus**
this should fix #988, but please test before merge.

### CODE REVIEW

**Comment by andlaus** [comment]
jenkins run norne_ebos please

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 1]
```diff
@@ -992,29 +993,41 @@ namespace Opm {
...
+
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+
+                const unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+                const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * ebosSimulator_.problem().porosity(cellIdx);
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: These are a lot of new lines of code. As far as I can judge the only change is that you now calculate the pore volume from ebos properties instead of taking the one stored in the properties. Is this the fix, i.e. are the values in `geo_` wrong. If not, what is the reason to obfuscate it with DUNE/ebos lingo and why is it even part of this PR?

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 2]
```diff
@@ -1056,25 +1069,31 @@ namespace Opm {
...
-                        hcpv[region] += pv[c] * hydrocarbon;
-                        pres[region] += pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value();
+                        const double pv =
+                            ebosSimulator_.model().dofTotalVolume(c)
+                            * ebosSimulator_.problem().porosity(c);
+                        hcpv[region] += pv * hydrocarbon;
+                        pres[region] += pv * fs.pressure(FluidSystem::oilPhaseIdx).value();
                     }
                 }
                 for (int c = 0; c < nc; ++c) {
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: Here you still use the old loop style and the obfuscated one from above. Why? Seems inconsistent and will raise questions in the future.

Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: BTW and totally unrelated. It seem we have still not fixed all region summations in parallel runs (e.g. by masking or as you did above by only computing on InteriorNodes. But that fix should be in a separate PR.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 3]
```diff
@@ -1136,15 +1158,18 @@ namespace Opm {
...
                     const int region = fipnum[c] - 1;
                     if (region != -1 && mask[c]) {
-                        fip_.fip[FIPData::FIP_PV][c] = pv[c];
                         const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(c, /*timeIdx=*/0);
                         const auto& fs = intQuants.fluidState();
                         const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
+                        const double pv =
+                            ebosSimulator_.model().dofTotalVolume(c)
+                            * ebosSimulator_.problem().porosity(c);
+                        fip_.fip[FIPData::FIP_PV][c] = pv;
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: We have calculated the pore volume three times now. I guess, each time we need to access two arrays now instead of one previously?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 4]
```diff
@@ -157,24 +157,31 @@ class SimulatorFullyImplicitBlackoilEbos
...
         std::ofstream tstep_os(tstep_filename.c_str());
 
+        const auto& schedule = eclState().getSchedule();
+
         // adaptive time stepping
         std::unique_ptr< AdaptiveTimeStepping > adaptiveTimeStepping;
         if( param_.getDefault("timestep.adaptive", true ) )
         {
-            adaptiveTimeStepping.reset( new AdaptiveTimeStepping( param_, terminal_output_ ) );
+
```
Comment (blattms) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Hu? The FIP caculations are connected to the settings of the adaptive time stepper? That is quite a surprise. Please explain!

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 5]
```diff
@@ -157,24 +157,31 @@ class SimulatorFullyImplicitBlackoilEbos
...
             // -1 means that we'll take the last report step that was written
-            //const int desiredRestoreStep = param_.getDefault("restorestep", int(-1) );
+            const int desiredRestoreStep = param_.getDefault("restorestep", int(-1) );
 
-            //            output_writer_.restore( timer,
-            //                                    state,
-            //                                    prev_well_state,
-            //                                    restorefilename,
-            //                                    desiredRestoreStep );
+            output_writer_.restore( timer,
```
Comment (blattms) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: and to this place, too?

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 6]
```diff
@@ -195,7 +202,6 @@ class SimulatorFullyImplicitBlackoilEbos
             }
         }
         std::vector<std::vector<double>> OOIP;
-
```
Comment (blattms) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: Even empty lines seem to influence it?!

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp [hunk 7]
```diff
@@ -221,47 +227,40 @@ class SimulatorFullyImplicitBlackoilEbos
...
             const Wells* wells = wells_manager.c_wells();
             WellState well_state;
             well_state.init(wells, state, prev_well_state, props_.phaseUsage());
 
             // give the polymer and surfactant simulators the chance to do their stuff
             handleAdditionalWellInflow(timer, wells_manager, well_state, wells);
 
-            // Compute reservoir volumes for RESV controls.
-            computeRESV(timer.currentStepNum(), wells, state, well_state);
```
Comment (blattms) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: I am giving up...

**Comment by andlaus** [comment]
> Please remove all changes that have nothing to do with #988.

not a big deal. as far as I can see, you mean the synchronization with FlowMainBase (which diverged a bit due to missing merges since FlowMainEbos was introduced)? I'll do as @atgeirr prefers.

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 6 above)
Comment (andlaus) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: no, but I suppose that the empty line is not there there in the flow_legacy version ;)

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 4 above)
Comment (andlaus) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: flow_legacy does it this way

File: opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp (hunk 5 above)
Comment (andlaus) [inline opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp]: flow_legacy again

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 3 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: depends on what you mean with *calculated* since the `.model().dofTotalVolume(c)` and `problem().porosity(c)` methods are trivial. I'd say the pore volume was *accessed* three times here.

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 2 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: > Here you still use the old loop style and the obfuscated one from above. Why?

obfuscation lies in the eye of the beholder ;). why? because it means less lines of changed code (didn't somebody complain about this recently) and at this point, one can assume that the cache for intensive quantities is "hot". but you're probably right that it should be changed to a loop over the grid because as-is it does not work with an overlap. (the same is the case for flow_legacy as far as I can see.)

Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: > BTW and totally unrelated. It seem we have still not fixed all region summations in parallel runs (e.g. by masking or as you did above by only computing on InteriorNodes. But that fix should be in a separate PR.

that's quite possible. the (sequential) FIP numbers after this PR are consistent, with flow_legacy, though...

**Comment by andlaus** [comment]
ok, I've replaced all counting loops by loops over the grid. this has the advantages that the same mechanism to loop over all DOFs is used uniformly and that the parallel and non-parallel versions use the same code path. for Norne in sequential mode, the FIP numbers still match those of flow_legacy.

**Comment by atgeirr** [comment]
> not a big deal. as far as I can see, you mean the synchronization with FlowMainBase (which diverged a bit due to missing merges since FlowMainEbos was introduced)? I'll do as @atgeirr prefers.

I think combining the "anti-diverging of FlowMain*" with "fix the FIP stuff" has lead to confusion. I think I want it all... So the best would be to split it in two PRs, then it all should be much clearer (and get merged more quickly). Agree?

**Comment by andlaus** [comment]
agreed & PR split: #1026

**Comment by atgeirr** [comment]
So are the FIP changes still in this PR, and ready for review?

**Comment by andlaus** [comment]
> So are the FIP changes still in this PR, and ready for review?

yes, except for the change which is required to make #1026 not segfault anymore (and which leads to some rather simple merge conflicts with the present PR)

**Comment by andlaus** [comment]
with #1026 merged, I've rebased this on top of the latest master to solve the merge conflicts. I suppose it is clean enough for review now...

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 8]
```diff
@@ -1016,12 +1014,15 @@ namespace Opm {
                 const unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
                 const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * ebosSimulator_.problem().porosity(cellIdx);
```
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: BTW: if the rock compressibility can be considered by simply using `intQuants.porosity().value()`.

Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Starting over on this review, I may say things again that I or @blattms already have said. Bear over with me please...

Why not keep using geo_.poreVolume()?

Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Also: did we have a clear answer to the question of including variable pore volume (rock compressibility)? I assume including it here would make flow_ebos behave different from flow_legacy?

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 9]
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
...
-                // mask[c] is 1 if we need to compute something in parallel
-                const auto & pinfo =
-                        boost::any_cast<const ParallelISTLInformation&>(istlSolver().parallelInformation());
-                const auto& mask = pinfo.updateOwnerMask( fipnum );
 
-                auto comm = pinfo.communicator();
-                // Compute the global dims value and resize values accordingly.
-                dims = comm.max(dims);
-                values.resize(dims, std::vector<double>(FIPData::fipValues,0.0));
+            comm.sum(hcpv.data(), hcpv.size());
```
Comment (blattms) [inline opm/autodiff/BlackoilModelEbos.hpp]: Seems like these values have not even been initialized yet due to the changed order of the computations. That happens down in line 1088 ff.

Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: At this point, these vectors are just zeros, probably a bug.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 10]
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
...
-                if (active_[ Oil ] && active_[ Gas ]) {
-                    for (int c = 0; c < nc; ++c) {
-                        const int region = fipnum[c] - 1;
-                        if (region != -1 && mask[c]) {
-                            values[region][FIPData::FIP_DISSOLVED_GAS] += fip_.fip[FIPData::FIP_DISSOLVED_GAS][c];
-                            values[region][FIPData::FIP_VAPORIZED_OIL] += fip_.fip[FIPData::FIP_VAPORIZED_OIL][c];
-                        }
-                    }
-                }
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: I'll just refer to the comments I made before about this loop in a different location... Also, braces please (as in that other location...)

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 11]
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
...
-                // to sum up the vector entries instead of looping over the regions.
-                for(int reg=0; reg < dims; ++reg)
-                {
-                    comm.sum(values[reg].data(), values[reg].size());
+                    values[region][FIPData::FIP_PV] += fip_.fip[FIPData::FIP_PV][cellIdx];
+                    values[region][FIPData::FIP_WEIGHTED_PRESSURE] += fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][cellIdx];
                 }
-#else
-                // This should never happen!
-                OPM_THROW(std::logic_error, "HAVE_MPI should be defined if we are running in parallel");
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Why remove this?

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 12]
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
...
-                        const auto& fs = intQuants.fluidState();
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-
-                        if (hcpv[region] != 0) {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbon / hcpv[region];
-                        } else {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pres[region] / pv[c];
-                        }
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Is this call the only reason why you change from the old-style loop with `c` to the new with (essentially) `elemCtx`?

**Comment by atgeirr** [comment]
Seems that @blattms and I both looked at this in parallel, happily no deadlock ensued...

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 8 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: > Why not keep using geo_.poreVolume()?

because its quite easy to switch away from it, with `geo_` it is much more difficult to also include rock compressibility if this turns out to be the Right Thing (TM) to do, it makes the parallel code less explicit (IMO) because it needs use non-standard mechanisms, and it forces the legacy object to be kept around even if it is not used by anything else...

Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: > Also: did we have a clear answer to the question of including variable pore volume (rock compressibility)? I assume including it here would make flow_ebos behave different from flow_legacy?

the current version of the PR disregards rock compressibility to get a match with flow_legacy. Whether it should be included or not, I cannot really say...

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 9 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: yeah, quite possible. I did not think about whether the old version is semantically correct when refactoring it. ;)

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 10 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: yeah, I was lost in PRs here. fixed, thanks.

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 12 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: the reason is primary that the "new" style works unconditionally, i.e. it works regardless of whether the cached intensive quantities are present or not (or in fact regardless of whether the intensive quantities are cached in the first place)

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 11 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: it not necessary (the communicator is always used) and we would have to go through hoops to keep it (which is a weird thing to do for code which is never going to be executed IMO).

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 9 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: right. hopefully it's fixed now...

**Comment by blattms** [comment]
It would really help a lot to put the parallel stuff into a different PR!

**Comment by andlaus** [comment]
> It would really help a lot to put the parallel stuff into a different PR!

you mean removing the calls to `comm`? that's easy but IMO it does not really make sense...

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 8 above)
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: While I am perhaps clinging a bit close to the Eclipse-like way of thinking here, I think of (or have over time become conditioned to it) pore volume as consisting of a reference pore volume and a dynamic multiplier (pressure dependence). And in computing the reference pore volume there are lots of nitty-gritty details, that are not reflected in the apparent formula used here (volume times porosity). If it is correct for this particular purpose, then it must be because either the volume or the porosity must include such things as NTG. I assume it is baked into the porosity. In that case it will be wrong for various polymer simulations for example, which include an adsorption term that depends on porosity.

To sum up, I really don't like the explicit use of a formula here to compute pore volume, where before we used an already-digested number.

Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: > And in computing the reference pore volume there are lots of nitty-gritty details, that are not reflected in the apparent formula used here (volume times porosity)

these nitty-gritty (NTG, MULTPV, MULTREGP, PORV, etc.) details are actually considered by the way the porosity is computed by EclProblem: https://github.com/OPM/ewoms/blob/master/ebos/eclproblem.hh#L982

As noted in OPM/ewoms#130, this is not very transparent and depends on lot of side conditions (mainly the ECL specification is to blame here, IMO) and on some details of the grid (e.g. how it implements MINPV).

>  I assume it is baked into the porosity.

exactly. I'm quick to admit that this is not exactly pretty, though.

> To sum up, I really don't like the explicit use of a formula here to compute pore volume, where before we used an already-digested number.

the reason why ebos uses porosity instead of pore volume is that the models in eWoms are not specific to the finite volume discretization and concept of a "pore volume" does not make any sense for discretizations that require the problem to be defined in terms of intensive quantities (e.g., finite element methods).

**Comment by atgeirr** [comment]
OK, it is now time to resurrect this work and restart the discussion on it. I am not sure if it has been affected significantly by recent development, but I'd ask you to rebase it in any case to make it green again. Sorry for letting it lie still for so long...

**Comment by andlaus** [comment]
ok, I've rebased this on top of the latest master and fixed the conflicts. as before rebasing the FIP numbers for flow ebos are much closer to those of flow_legacy with this PR but they are not quite a perfect match...

**Comment by andlaus** [comment]
I've found a small bug in the calculation of the initial fluid in place. now the numbers are *very* close to those produced by ECL...

**Comment by andlaus** [comment]
jenkins run norne_ebos please

**Comment by atgeirr** [comment]
> review is stale and most issues have been addressed.

In general I think the PR author should not dismiss reviews... Although you are correct that its primary concerns have been addressed.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 13]
```diff
@@ -1029,12 +1027,15 @@ namespace Opm {
                 const unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
                 const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * intQuants.porosity().value();
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Please add comment explaining why this is correct. Important points:
 - the intQuants.porosity() can be outside [0,1] due to pore volume manipulation
 - while the problem.porosity() would also account for this manipulation, only the intQuants version includes the pressure-dependent porosity multiplier

**Comment by andlaus** [comment]
> In general I think the PR author should not dismiss reviews... Although you are correct that its primary concerns have been addressed.

yeah, I know and usually I don't. my main reason to do this was that it seems like @blattms dropped out of that discussion about three weeks ago.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 14]
```diff
@@ -1045,150 +1046,100 @@ namespace Opm {
...
 
-                //Accumulate phases for each region
-                for (int phase = 0; phase < maxnp; ++phase) {
-                    for (int c = 0; c < nc; ++c) {
-                        const int region = fipnum[c] - 1;
-                        if (region != -1 && mask[c]) {
-                            values[region][phase] += fip_.fip[phase][c];
-                        }
-                    }
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: Please keep the loop iterators in local scope.

File: opm/autodiff/BlackoilModelEbos.hpp [hunk 15]
```diff
@@ -1045,150 +1046,100 @@ namespace Opm {
...
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
 
-                        values[region][FIPDataType::FIP_PV] += fip_.fip[FIPDataType::FIP_PV][c];
-                        values[region][FIPDataType::FIP_WEIGHTED_PRESSURE] += fip_.fip[FIPDataType::FIP_WEIGHTED_PRESSURE][c];
+                unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+                const int region = fipnum[cellIdx] - 1;
+                if (region != -1) {
+                    const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+                    const auto& fs = intQuants.fluidState();
+                    const double pv =
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelEbos.hpp]: This formula is repeated very often, did you consider storing the value?

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 13 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: comment added. IMO the second point is self-evident because problem.porosity() returns a scalar instead of an evaluation. (i.e., it cannot be solution dependent.)

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 14 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: do you mean to *create* a local scope around each of these loops to achive this?

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 15 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: it is just a single multiplication, so storing it somewhere is not worth the hassle IMO. (also, it would slightly increase memory consumption and it may even be slightly slower due to cache effects; Yeah, we don't need to care about memory consumption and this part of the code is not very performance critical...)

File: opm/autodiff/BlackoilModelEbos.hpp (hunk 14 above)
Comment (andlaus) [inline opm/autodiff/BlackoilModelEbos.hpp]: ok: I've changed this so that the element iterators are created in the first part of the `for` statements. the result is harder to read IMO, so I would not mind if you requested to revert that patch before the PR gets merged.

**Comment by atgeirr** [comment]
About the scoping: I think it's more readable now... Beauty is in the eye of the beholder! Intend to merge when Travis is green (don't think we need to rerun Jenkins).

**Comment by andlaus** [comment]
> About the scoping: I think it's more readable now... Beauty is in the eye of the beholder!

ok, that's fine with me. (I don't care too much about it.)

> (don't think we need to rerun Jenkins).

I don't think so either; as far as I can see, nothing has changed functionality-wise since the last jenkins run.

**Comment by andlaus** [comment]
that PR was a long story. thanks for merging :)
